pkg/controller/atlasdatafederation/datafederation_controller.go

```diff
@@ -0,0 +1,273 @@
+		ctx.SetConditionFromResult(status.DataFederationPEReadyType, result)
+		return result.ReconcileResult(), nil
+	}
+
+	if result = r.ensureConnectionSecrets(ctx, project, dataFederation); !result.IsOk() {
+		return result.ReconcileResult(), nil
+	}
+
+	if dataFederation.GetDeletionTimestamp().IsZero() {
+		if !customresource.HaveFinalizer(dataFederation, customresource.FinalizerLabel) {
+			err = r.Client.Get(contextInt, kube.ObjectKeyFromObject(dataFederation), dataFederation)
+			if err != nil {
+				result = workflow.Terminate(workflow.Internal, err.Error())
+				return result.ReconcileResult(), nil
+			}
+			customresource.SetFinalizer(dataFederation, customresource.FinalizerLabel)
+			if err = r.Client.Update(contextInt, dataFederation); err != nil {
+				result = workflow.Terminate(workflow.Internal, err.Error())
+				log.Errorw("failed to add finalizer", "error", err)
+				return result.ReconcileResult(), nil
+			}
+		}
+	}
+
+	if !dataFederation.GetDeletionTimestamp().IsZero() {
+		if customresource.HaveFinalizer(dataFederation, customresource.FinalizerLabel) {
+			if customresource.ResourceShouldBeLeftInAtlas(dataFederation) {
+				log.Infof("Not removing AtlasDataFederation from Atlas as the '%s' annotation is set", customresource.ResourcePolicyAnnotation)
+			} else {
+				if err = r.deleteDataFederationFromAtlas(contextInt, dataFederation, project, log); err != nil {
+					log.Errorf("failed to remove DataFederation from Atlas: %s", err)
+					result = workflow.Terminate(workflow.Internal, err.Error())
+					ctx.SetConditionFromResult(status.DataFederationReadyType, result)
+					return result.ReconcileResult(), nil
+				}
+			}
+			if err = r.removeDeletionFinalizer(contextInt, dataFederation); err != nil {
+				result = workflow.Terminate(workflow.Internal, err.Error())
+				log.Errorw("failed to remove finalizer", "error", err)
+				return result.ReconcileResult(), nil
+			}
+		} else {
+			return result.ReconcileResult(), nil
+		}
+	}
+
+	ctx.SetConditionTrue(status.ReadyType)
+	return workflow.OK().ReconcileResult(), nil
+}
+
+func (r *AtlasDataFederationReconciler) removeDeletionFinalizer(context context.Context, df *mdbv1.AtlasDataFederation) error {
+	err := r.Client.Get(context, kube.ObjectKeyFromObject(df), df)
+	if err != nil {
+		return fmt.Errorf("cannot get AtlasDeployment while adding finalizer: %w", err)
+	}
+
+	customresource.UnsetFinalizer(df, customresource.FinalizerLabel)
+	if err = r.Client.Update(context, df); err != nil {
+		return fmt.Errorf("failed to remove deletion finalizer from %s: %w", df.GetName(), err)
+	}
+	return nil
+}
+
+func (r *AtlasDataFederationReconciler) deleteDataFederationFromAtlas(ctx context.Context, df *mdbv1.AtlasDataFederation, project *mdbv1.AtlasProject, log *zap.SugaredLogger) error {
+	log.Infof("Deleting DataFederation instance: %s from Atlas", df.Spec.Name)
+	connection, err := atlas.ReadConnection(log, r.Client, r.GlobalAPISecret, project.ConnectionSecretObjectKey())
```

Comment:
Why do we need to recreate the client if it's available in the caller and can be passed to this function?

Reply:
Good point, fixed!